cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        LOG.error("No output from cmd to show the user profile dir.")
+        return None
+
+    return win_path_2_wsl(home.rstrip())
+
+
+def machine_id():
+    """
+    Returns the local machine ID value from /etc/machine-id.
+    """
+    MACHINE_ID_FILE = "/etc/machine-id"
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
+        LOG.debug("%s file not found", MACHINE_ID_FILE)
+        return None
+
+    return util.load_file(MACHINE_ID_FILE, decode=True)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    lsb_rel = util.lsb_release()
+    distribution_id = lsb_rel["id"]
+    release_codename = lsb_rel["codename"]
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "config.user-data",
+    ]
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_LOCAL
+        self.distro = distro
+        self.instance_name = instance_name()
+
+    def find_user_data_file(self) -> Optional[PurePath]:
+        """
+        Finds the most precendent of the candidate files that may contain
+        user-data, if any, or None otherwise.
+        """
+        profile_dir = win_user_profile_dir()
+        if profile_dir is None:
+            LOG.warning(
+                "Cannot proceed without finding the Windows %USERPROFILE% dir."
+            )
+            return None
+
+        seed_dir = os.path.join(profile_dir, ".cloud-init")
+        if not os.path.isdir(seed_dir):
+            LOG.warning("%s directory doesn't exist.", seed_dir)
+            return None
+
+        # Notice that file name casing is irrelevant here. Windows filenames
+        # are case insensitive. Even though accessed through Linux, path
+        # translation just works with whichever casing we try.
+        for filename in candidate_user_data_file_names(self.instance_name):
+            file = os.path.join(seed_dir, filename)
+            if os.path.isfile(file):
+                return PurePath(file)
+
+        LOG.warning(
+            "%s doesn't contain any of the expected user-data files", seed_dir
+        )
```

Comment:
Since we require that at least one user-data file exists which meets the launch requirements for this instance, this should be an error, and the error message should probably state that a user-data file was that isn't present is something like
`%USERPROFILE%\.cloud-init\config.user-data`.  

Most clouds support an empty user-data file and cloud-init handles that fine because user-data content is optional, but the structure of valid metadata service configuration requires that we at least 

If possible, ./tools/ds-identify should probably also validate that a top-level `%USERPROFILE%\.cloud-init\config.user-data`  exists on the system if we can check that early enough in boot. Without it, we can't officially detect WSL datasource with the python get_data call.

Reply:
Instead of relying on an always present user data file, we could just assign the empty string to `self.userdata_raw` if a file is not found. Wouldn't that have the same effect?

Since the datasource only craws user data, if there is none, there is nothing for this datasource to do. Ideally we should really do nothing, not even DataSourceNone should play any role under such situation. Is there a good way to create that behavior?

Reply:
> Instead of relying on an always present user data file, we could just assign the empty string to self.userdata_raw if a file is not found. Wouldn't that have the same effect?

Sorry I missed this response. I think it is a good constraint/restriction to place on the WSL datasource to only activate/discover if it sees some applicable user-data file. This would allow customizatation of some environments that may the ability to deactivate cloud-init on a WSL image stream which happens to have cloud-init installed (and configured to potentially use the WSL datasource). This would then behave like our nocloud datasource, if some expected NoCloud user-data file doesn't exist in /var/lib/cloud/seed/nocloud/user-data, then cloud-init won't try to discover that datasource and configure the machine. There may be use-cases we haven't thought of where external to the WSL image we want a mechanism to leave cloud-init inert, and the absence of any applicable user-data file would be the easiest knob to expose to the host windows system/environment.

Reply:
Gotcha! Unfortunately we'll need to keep some logic duplicated in between ds-identify and the datasource until all distros support the first, but that's fine.